File: dart/lib/output.dart
import 'dart:convert';
import 'dart:io';
import 'benchmark_result.dart';

class SystemInfo {
  final String os;
  final String arch;
  final int cpus;

  SystemInfo({
    required this.os,
    required this.arch,
    required this.cpus,
  });

  Map<String, dynamic> toJson() => {
        'os': os,
        'arch': arch,
        'cpus': cpus,
      };
}

class BenchmarkReport {
  final String language;
  final String timestamp;
  final String dartVersion;
  final SystemInfo system;
  final List<BenchmarkResult> tests;
  final double totalTimeSeconds;

  BenchmarkReport({
    required this.language,
    required this.timestamp,
    required this.dartVersion,
    required this.system,
    required this.tests,
    required this.totalTimeSeconds,
  });

  Map<String, dynamic> toJson() => {
        'language': language,
        'timestamp': timestamp,
        'dart_version': dartVersion,
        'system': system.toJson(),
        'tests': tests.map((t) => t.toJson()).toList(),
        'total_time_seconds': totalTimeSeconds,
      };
}

class Output {
  static void printResults(List<BenchmarkResult> results) {
    print('\n=== BENCHMARK RESULTS ===');
    for (final result in results) {
      print('Test: ${result.test}');
      print('  Duration: ${result.durationNs} ns');
      print('  Memory: ${result.memoryBytes} bytes');
      print('  Operations: ${result.operations}');
      print('  Ops/sec: ${result.opsPerSec.toStringAsFixed(2)}');
      print('');
    }
  }

  static Future<void> saveResultsToJSON(
      List<BenchmarkResult> results, double totalTime) async {
    final report = BenchmarkReport(
      language: 'dart',
      timestamp: DateTime.now().toIso8601String(),
      dartVersion: Platform.version.split(' ').first,
      system: SystemInfo(
        os: _getOSName(),
        arch: _getArchitecture(),
        cpus: Platform.numberOfProcessors,
      ),
      tests: results,
      totalTimeSeconds: totalTime,
    );

    final now = DateTime.now();
    final timestamp = '${now.year.toString().padLeft(4, '0')}'
        '${now.month.toString().padLeft(2, '0')}'
        '${now.day.toString().padLeft(2, '0')}_'
        '${now.hour.toString().padLeft(2, '0')}'
        '${now.minute.toString().padLeft(2, '0')}'
        '${now.second.toString().padLeft(2, '0')}';
    final filename = 'benchmark_dart_$timestamp.json';

    final encoder = JsonEncoder.withIndent('  ');
    final jsonString = encoder.convert(report.toJson());

    final file = File(filename);
    await file.writeAsString(jsonString);
    print('Results saved to: $filename');
  }

  static String _getOSName() {
    if (Platform.isWindows) return 'Windows';
    if (Platform.isMacOS) return 'macOS';
    if (Platform.isLinux) return 'Linux';
    if (Platform.isAndroid) return 'Android';
    if (Platform.isIOS) return 'iOS';
    return 'Unknown';
  }

  static String _getArchitecture() {
    final executable = Platform.executable;
    if (executable.contains('arm64')) return 'arm64';
    if (executable.contains('x64') || executable.contains('x86_64')) return 'x86_64';
    if (executable.contains('arm')) return 'arm';
    return 'unknown';
  }
}

File: julia/src/output.jl
using JSON3
using Dates

include("benchmark_result.jl")

struct SystemInfo
    os::String
    arch::String
    cpus::Int
end

struct BenchmarkReport
    language::String
    timestamp::String
    julia_version::String
    system::SystemInfo
    tests::Vector{BenchmarkResult}
    total_time_seconds::Float64
end

function print_results(results::Vector{BenchmarkResult})
    println("\n=== BENCHMARK RESULTS ===")
    for result in results
        println("Test: $(result.test)")
        println("  Duration: $(result.duration_ns) ns")
        println("  Memory: $(result.memory_bytes) bytes")
        println("  Operations: $(result.operations)")
        println("  Ops/sec: $(round(result.ops_per_sec, digits=2))")
        println()
    end
end

function save_results_to_json(results::Vector{BenchmarkResult}, total_time::Float64)
    report = BenchmarkReport(
        "julia",
        Dates.format(now(UTC), "yyyy-mm-ddTHH:MM:SSZ"),
        string(VERSION),
        SystemInfo(
            get_os_name(),
            get_architecture(),
            Sys.CPU_THREADS
        ),
        results,
        total_time
    )

    timestamp = Dates.format(now(), "yyyymmdd_HHMMSS")
    filename = "benchmark_julia_$(timestamp).json"

    # BenchmarkResultを辞書に変換する関数
    function benchmark_result_to_dict(br::BenchmarkResult)
        return Dict(
            "test" => br.test,
            "duration_ns" => br.duration_ns,
            "memory_bytes" => br.memory_bytes,
            "operations" => br.operations,
            "ops_per_sec" => br.ops_per_sec
        )
    end

    # BenchmarkReportを辞書に変換
    report_dict = Dict(
        "language" => report.language,
        "timestamp" => report.timestamp,
        "julia_version" => report.julia_version,
        "system" => Dict(
            "os" => report.system.os,
            "arch" => report.system.arch,
            "cpus" => report.system.cpus
        ),
        "tests" => [benchmark_result_to_dict(result) for result in report.tests],
        "total_time_seconds" => report.total_time_seconds
    )

    open(filename, "w") do io
        JSON3.pretty(io, report_dict)
    end

    println("Results saved to: $filename")
end

function get_os_name()
    if Sys.iswindows()
        return "Windows"
    elseif Sys.isapple()
        return "macOS"
    elseif Sys.islinux()
        return "Linux"
    else
        return "Unknown"
    end
end

function get_architecture()
    arch = string(Sys.ARCH)
    if arch == "x86_64"
        return "x86_64"
    elseif arch == "aarch64"
        return "arm64"
    else
        return arch
    end
end